### PR TITLE
Add alternative memory-optimized code

### PR DESCRIPTION
**Pull request by rasbt**
In GRPO, the standard implementation used in the book currently seems to use a growing amount of memory over time. 
E.g., after 5000 GRPO steps the GPU memory grows to 70 GB.

I am currently unsure whether that's 

a) just normal, because in the beginning, we don't have all the rollouts using the max token length, and at some point later in training the rollouts all use the max token lengths
b) due to stray or cached tensors due to `torch.cat`
c) something else

This branch implements an alternative version that

- uses pre-allocated attention masks of a fixed size instead of generating them on the fly
- uses pre-allocated KV cache buffers instead of growing the KV cache on the fly with `torch.cat`

Note that it looks like there are a lot of changes in the file diff. But that's just because I added the modified files as new files under `reasoning_from_scratch/memopt/`.

To see the leaner, actual file diff, check out the [changes](https://github.com/rasbt/reasoning-from-scratch/pull/115/changes) in #115

<a href="https://github.com/rasbt/reasoning-from-scratch/pull/115/changes">Screenshot 2026-01-10 at 11 19 45 AM</a>

### CODE REVIEW

**Comment by rasbt** [comment]
The conclusion here is that there is no memory leak and everything appears to be fine. The alternative version looks more efficient, but in practice it also uses about the same memory (68 instead of 71 GB), and tokens/sec wise both implementations average around 51 tokens/sec on a DGX Spark. So I am closing for now as there is no clear advantage.
